### ISSUE TITLE
Route/Controller API

### ISSUE THREAD

**Issue by jhuntoo** [CitadelJS/core#10]
I'd imagine much off https://github.com/inversify/inversify-express-utils will be straight copy

We should compare with https://github.com/pleerock/routing-controllers

& https://github.com/DemgelOpenSource/demgel-mvc to see how we can design the best API possible.

I'll take a look soon, but feel free to comment!

**Comment by codyjs** [comment]
Things we get from inversify-express-utils: 

* Controller class decorator
* HTTP verb method decorators

Things that would also be cool to have:

* RouteParam/Body/Query parameter decorators
* Request/Response parameter decorators

One thing to think about: Will it be required to return something from a handler method? If not, will we provide a Response.send() method or equivalent?

**Comment by otbe** [comment]
I think we should copy most of the API, like ```@Controller @Get ...``` and add some things like already mentioned by @codyjs. 
But we should carefully rethink the implementation on some edges :)

> One thing to think about: Will it be required to return something from a handler method? If not, will we provide a Response.send() method or equivalent?

I think its not required to return something besides a valid status code. Something like
```javascript
// 200 is default response code if no exception occurs
@Post('/users/:id/activate')
async activate(id): Promise<void> {
   await activateUser(id);
}
```
should be possible. If nothing is retuned (undefined|null) we can fulfill the response by calling ```res.end()``` (or equivalent) internally. Maybe we should add ```@Produces('application/json')``` or something like that and force a valid response, which could be ```{}``` in case of undefined or null. I don't know if this is necessary. express ```res.send()``` does a lot of magic to determine ContentType and needed conversions (for example ```res.send({})``` will execute JSON.stringify() internally and set ContentType to ```application/json```). Im not sure about that :/

**Comment by codyjs** [comment]
I agree, returning shouldn't be required.

>  Maybe we should add @Produces('application/json') or something like that and force a valid response, which could be {} in case of undefined or null.

Wouldn't it be valid to simply send an empty response body for undefined or null? If it's required to return an empty object, then I think it should be up to the user to `return {};`

I think we should rely as much as possible on the underlying framework (express/koa/etc) to handle ContentType and such. In this case, assuming express is used, if we detect an Object was returned, we can pass it to `res.send()`.

**Comment by otbe** [comment]
> Wouldn't it be valid to simply send an empty response body for undefined or null?

That depends a lot on status code, ContentType and how restrictive we want to be. For example: if we add ```@Produce('application/json')``` or the dev set it by calling ```this.setContentType()``` and the response header is probably set some rest clients will raise an error if the API returns 200 without body (by simply call  JSON.parse(null)). I wonder if we should go this way or make devs responsible for such things.

Btw. I would love to see ```@Produces()``` instead of ```this.setContentType()``` (or something like that). Its more declarative and it forces devs to write good APIs (and not return different ContentType based on business logic). :)

**Comment by codyjs** [comment]
I checked in express, it seems that it is smart enough to remove the ContentType header when the body is empty. This is also the behavior in Spring MVC, even in a controller where the ContentType was specified as `application/json`. 

So we may just need some logic that removes that header if the result is undefined or null.

+1 for `@Produces()` and declarative syntax!

**Comment by lholznagel** [comment]
I like the `@Produces` reminds me a bit on spring boot.

**Comment by codyjs** [comment]
@tenowg That's basically what we want. Personally I would like to have the param injection API be similar to https://github.com/pleerock/routing-controllers#inject-parameters

**Comment by tenowg** [comment]
@codyjs agreed, I assumed in my implementation that the code would list the params in the order listed, which could lead to issues. So a Params decorator would be very useful, and really not hard to do.

**Comment by jhuntoo** [comment]
@Produces 👍, on method &controller, maybe support some way of setting it globally as well.
 
Agree with @codyjs, routing-controllers's style appeals to  me better for query and route params.

**Comment by tenowg** [comment]
I am not feeling the @Produces as much as I would like to see a Result returned. The result would decide what the header would be. Thus a developer could return a JsonResult or a ViewResult, a JsonResult is handled and returns a `application/json` while a ViewResult would use (in the case of express) res.render("..."); or a FileResult would use res.sendFile("...")... and all could be returned from a single handler without the need to declare.

**Comment by otbe** [comment]
@tenowg in my opinion returning different ContentTypes from **one** handler based on business logic is exactly what we should be prevent for the most cases. Its hard to test and error prone. 

If we need multiples ContentTypes from the same route we can do something like [this](https://gist.github.com/otbe/322966135debb7084eaf12994afee1a8). So ```Accept``` header will decide which handler will be triggered. Browsers takes the first one and Rest clients takes the latter one. Personally I don't like this style and in my opinion its not very common these days, because most apps will have a Rest-ish API and deliver html, js and other assets via a static file server. But thats another story so far.

But you are right there are some use cases for "dynamic" ContentTypes. For example dynamic generated files (image/jpeg, application/pdf) and a download endpoint for arbitrary content at all. For this special case we can do something like [this](https://gist.github.com/otbe/7b585ab92ca02f0bc504e6db495e990e).

Let me be more explicit. What I don't want to allow is:
```javascript
@Get('/')
async whatEverYouWant(): Promise<Resultish> {
   let day = await computeDayOfWeek();
   if(day === 'friday') {
      return JsonResult({ tgif: true });
   } else if(day === 'monday') {
      return ViewResult('wtf.html');
   } else {
      return new Result('NobodyWorksHere', 'application/pdf');
   }
}
```
😁 

What do you guys think about it?

**Comment by codyjs** [comment]
@otbe Agreed that your methods (which should correspond to a route) should only return one type of thing.

@tenowg Maybe I'm wrong, but I think you might be concerned with `@Produces()` on the Controller itself, which would mean all the methods in the Controller would be locked into one ContentType. But it sounds like you could still mix different ContentTypes within the same controller if you use `@Produces()` on the method level.

```ts
@Get('/json')
@Produces('application/json')
async dayJson(): Promise<any> {
   let day = await computeDayOfWeek();
   return JsonResult({ tgif: day === 'friday' });
}

@Get('/html')
@Produces('text/html')
async dayHtml(): Promise<any> {
   let day = await computeDayOfWeek();
   return ViewResult('wtf.html');
}
```

**Comment by tenowg** [comment]
@otbe not disagreeing that dynamic content-types should be discouraged, there may be a way to discourage this, but I think we are going to have enough annotations to start with, and adding a another just to decide return type is unneeded and redundant... in most cases express (if we use that) decides it for you. And we shouldn't really decide what a developer wants to do... only put in Best Practices.

Would like to avoid:

``` typescript
@Get('/json')
@Authorize("Admin")
@Parameters(blah)
@AnotherAnnotation()
@Produces()
@AnotherMiddleware()
async method(): Promise<JsonResult> {
}
```

Which leads to a kinda hell in itself. Would also like to discourage the use of the response object directly, which is something we already discussed, so returning a defined Result Typed Object would make developers life's easier.

Although I do agree that using the Produces annotation could be useful in situations where you need to define what Content-type is produced, wouldn't want to reuse code just to send a Json/Xml response, which would be, in this specific case, fixed by returning either a JsonResult or a XmlResult... but yes it would also allow the developer to return a ViewResult, but that would be on him.

Maybe we could consider adding this to the Method Decorator @Get("/json", "application/json")

**Comment by lholznagel** [comment]
@tenowg good argument.
We could overload the @Get decorator
@Get(route) -> Route is always required
@Get(route, produces) -> Optional produces
or what's maybe better and gives us more freedom
@Get(route, options) -> Optional options where things like produces and some other options can be set

**Comment by remojansen** [comment]
Too many annotations looks bad but a really long annotation look bad as well.

👍 to

 >  @Get("/json", "application/json")

I would suggest to do like in ASP and when a annotation repeats allow to be used at controller level so instad of:

```ts
class MyController {
 
    @Get("/json", "application/json")
    @Authorize("Admin")
    @CacheResponse()
    @Cors({ origins: "http://example.com" })
    async method1(): Promise<JsonResult> {
        // ...
    }

    @Get("/json2", "application/json")
    @Authorize("Admin")
    @Cors({ origins: "http://example.com" })
    async method2(): Promise<JsonResult> {
        // ...
    }

}
```

We can write:

```ts
@Authorize("Admin")
@Cors({ origins: "http://example.com" })
class MyController {
 
    @Get("/json", "application/json")
    @CacheResponse()
    async method1(): Promise<JsonResult> {
        // ...
    }

    @Get("/json2", "application/json")
    async method2(): Promise<JsonResult> {
        // ...
    }

}
```

**Comment by otbe** [comment]
@remojansen oh I thought this way of implementing it was common sense :)
Should be harder to not implement it in this way.

**Comment by jhuntoo** [comment]
And for those who only ever want application/json

```
const server = new Citadel(expressEngine());
server.configure( { defaultContentType = 'application/json' });
``
This could be overridden on the controller, method decorator,  or method response.

Order of precedence for configuration would obviously be
1. Global config
2. Controller config
3. Route config
4. Response object

This would apply to a lot of things like, content-type, other HTTTP headers, Authentication & Authorization

**Comment by lholznagel** [comment]
Would also say that there is a default config. A little bit unnecessary to write it again and again if its not changing.

**Comment by jhuntoo** [comment]
Greed. Personally I like a framework that just works OOTB, and don't have any problems with the framework making some sensible default configuration decisions. ie Convention over configuration.

One thing that  kinda annoys me about existing frameworks, is that you have to provide a body parser, to do *anything* meaningful. And there tons of options out there for parsers, but ultimately i don't care, i just want anyone that works. And a parser seems like only a small piece of code, to not include by default.

That said, I've ever only written JSON API's for SPA's, so I"m only seeing it from one angle.

**Comment by otbe** [comment]
@jhuntoo 👍
Im sitting in the same boat.

Thats exactly what ```@Produces```/ ```@Consumes``` can deliver. 
- ```@Consumes``` checks if expected ContentType matches the ContentType header in request and transforms the body (JSON.parse)
- ```@Produces``` checks if ContentType matches the Accept request header and transforms the response (JSON.stringify)

If no match occurs the request will be passed to the next route (or generate an error if no other handler is available). There might be a difference between Controller scoped and method scoped Producer/Consumes, but thats an implementation detail.

**Comment by remojansen** [comment]
I have changed my mind...

I like the following:

```ts
@Get("/json")
@Produces("application/json")
@Consume("application/json")
```
More than something like:

```ts
@Get("/json", "application/json", "application/json")
```

The first option is more declarative and I understand what it does without reading any docs.

Another option could be:

```ts
@Get("/json", { consumes: "application/json", produces: "application/json" })
```

But i don't like it is too verbose...

**Comment by jhuntoo** [comment]
@remojansen 

With the aim of reducing configuration. I'd like to see this
``` typescript
@JsonController()
class MyController {

  @Get("/foo")
  async getFoo(): Promise<Response> {
    // ...
  }
}
```
which is syntactic sugar for 
``` typescript
@Controller({ consumes: "application/json", produces: "application/json" }) //
class MyController {

  @Get("/foo")
  async getFoo(): Promise<Response> {
    // ...
  }
}
```
of course you can override for the exception cases like this...
``` typescript
@JsonController()
class MyController {

  @Get("/pdf", {  produces: "application/pdf" }) // consumes: "application/json" is all ready set from controller
  async getFoo(): Promise<Response> {
    // ...
  }

  @Get("/jsonbydefault") // consumes and produces by default
  async getFoo(): Promise<Response> {
    // ...
  }
}
```

**Comment by pleerock** [comment]
what you guys discussing already implemented in [routing-controllers](https://github.com/pleerock/routing-controllers) and even more with best practices applied.

so I have several questions:
1. why to create another bicycle? 
2. why not to contribute to improve already exist things, that are opened to improvements
3. why not to extend already exist thing to create something new on top of it (but its not a case, what I see from the discussion)

**Comment by jhuntoo** [comment]
Fair questions - I actually like your approach on many aspects, and indeed I imagine we can lift much of your hard work into this project. Which is why I included you into the discussion over on inversify.

1. We are not creating a new bicycles, we are taking stock of all the existing bicycle parts that exist and putting together what looks good and writing bits that we, as a group think will be better for the most people possible
2. I feel that a new org, with module repos, is the best way forward because we can build a "name", and more importantly to promote cohesiveness, and coherence, so code bases are built to the same standards and similiar processes. Rather than importing stuff from pleerock/routing-controllers , somerepo/anothercomponent. We want modules to be covered by integration tests across all repos for correctness, so that we can say that @citadel/core @citadel/routes @citadel/express actually work together now on master. This is much easier and quick to resolve failures if one community handles it, as opposed to individuals with less capacity for this work.
3. much the same as 2.

In summary, I personally want to use a lot of your existing work (route decorators and engines at least!), and would love for you to buy into this project, and move it forward together. However i think it makes sense to review each piece, as we go, and build a testing framework around what we release for 

My question for you is ? What is stopping you from becoming a contributor to a community built Citadel project ?

**Comment by pleerock** [comment]
> importing stuff from pleerock/routing-controllers

Yeah importing will look like `import * from "routing-controllers"` and it should not be issue even for naming. Don't think that "name" is so important in **imports**, because imports are being imported automatically and collapsed in most IDEs. Having `@citadel` prefix in imports is just additional beauty, that is not always possible to follow. But anyway you can reexport them from your modules, so this should not be an issue.

> What is stopping you from becoming a contributor to a community built Citadel project ?

I'll support and help to deal with any issue community (citadel project or any other project that is using components I have created). And this will be my contribution to Citadel project.

I developed this library over one year ago and using it in production since that. I developed it by myself only because there wasn't any alternatives, otherwise I would used existing libraries.

**Comment by remojansen** [comment]
For me providing all elements as a framework is a big thing:`@citadel/core` `@citadel/routes` `@citadel/express` because allows us to have a better control over releases and a more cohesive experience.

At the same time I see the point of @pleerock and one option could be to develop a project named `@citadel/routes` and this module could use `routing-controllers` under the hood. 

However, for this to happen I would demand a higher level of security from `routing-controllers`:

- The project needs a higher test coverage
- CI
- CD
- PR Reviews just like we plan to do with our components. 

Without this level of security I would prefer to write the module from scratch.

If we use `routing-controllers` to copy code or as inspiration we should give credit to the project.

Open source is about helping each other to progress so I would be happy to contribute to`routing-controllers` and be helped by  the `routing-controllers` team.

**Comment by jhuntoo** [comment]
All good points! Let's help each other out where possible.

**Comment by pleerock** [comment]
totally agree with ** higher level of security** point - this is in my todo list. Just need more motivation and time.  But now if there will be some plan I can push harder on this side and with help of you guys we can make routing-controllers part of citadel
